### ISSUE TITLE
Support DKIM in TIdSMTPClient

### ISSUE THREAD

**Issue by chuacw** [IndySockets/Indy#304]
Provide DKIM support in TIdSMTPClient.

**Comment by rlebeau** [comment]
The main hurdle to supporting DKIM is that it requires generating a digital signature and hash of an email's body and inserting those values into the email's headers, however TIdMessage dynamically generates an email's body as it is being sent/saved (to insert MIME boundaries, etc), so the body is not available in its entirety at the time the relevant DKIM headers would need to be generated.  This could be mitigated by first encoding the entire email body to a memory stream or a temp file, then hash/sign that before then sending/saving that data as-is.  However, this would be undesirable for emails with large bodies/attachments present.  So, need to find a way to handle this as efficiently as possible.

**Comment by chuacw** [comment]
This could be spun off in a thread to handle the calculations.

Since TIdMessage dynamically generates the email's body, you could also create a new event that fires before the email gets sent, so that anyone can listen on the event, generate the DKIM signature, and add it to TIdMessage, before it finally gets sent.

**Comment by chuacw** [comment]
Also, hashing/signing is on individual fields specified by the DKIM RFC specification, not hashing on the entire email.

**Comment by rlebeau** [comment]
I'm not concerned about the speed of signing, so I don't think a worker thread is needed.  I'm more concerned that because the email data is generated dynamically, it can **change** from the time it is generated for DKIM hashing/signing, and the time it is generated again for transmission.  Headers like "Date" can change value, but also body data like MIME boundaries can change, too.  Hashing/signing an email for DKIM does not involve headers only (and besides, the only header that the RFC requires is "From", the signer can chose other headers to use as well), but also involves body data, too (up to an octet length that the signer chooses, which _can_ be 0 but that is not desirable. By default, the entire body is hashed).  So, what gets sent out over the wire will have to match what is used for hashing/signing, or else the recipient won't be able to validate the email properly.  Adding new events before the transmission won't solve that issue.

**Comment by chuacw** [comment]
But if you add new events before the email is sent, someone else could then take a crack at implementing DKIM signing. ;o)

**Comment by rlebeau** [comment]
OK, but I don't really see the point of doing that, since THERE IS NO GUARANTEE that whatever they decide to sign will be what actually gets transmitted, thus invalidating the signature.

**Comment by chuacw** [comment]
You seem to be saying that TIdMessage at some point in time, before sending, generates the entire email.

I'm asking for an event to be generated before the email gets sent. 

You then say that there's no guarantee the data is what will be actually sent even if an event is generated so that the event can inspect TIdMessage for the contents to sign. Is there some nuance or subtlety I'm missing?

**Comment by rlebeau** [comment]
> You seem to be saying that TIdMessage at some point in time, before sending, generates the entire email.

No, not BEFORE it is sent - WHILE it is being sent.  The email content that is actually transmitted on the wire is generated dynamically as it is being transmitted.  `TIdMessage` does not save its content to a single blob and then that blob gets transmitted as-is.  `TIdMessageClient` takes _pieces_ of the `TIdMessage` at a time, formats each piece as needed, and transmits them as soon as they are formatted.  Headers like `Date` are updated dynamically before the send before.  But text parts and attachments are sent in a loop, formatted to bytes as they are being read from, with MIME boundaries generated and transmitted in between loop iterations.

So this makes it difficult to perform pre-send operations, like calculating a hash of the entire email, because the entire email doesn't exist yet in its final transmitted form at the time the hash is needed.

> I'm asking for an event to be generated before the email gets sent.

I know, and I'm saying that such an event WON'T help address this issue, because the final email that is transmitted DOESN'T EXIST YET at the time such an event would be triggered.

> You then say that there's no guarantee the data is what will be actually sent even if an event is generated so that the event can inspect TIdMessage for the contents to sign

Correct, because certain pieces of the email are generated dynamically at the moment they are actually being transmitted (MIME boundaries, for instance), or may change value dynamically (the `Date` header, etc).

> Is there some nuance or subtlety I'm missing?

Yes - how `TIdMessage` and `TIdMessageClient` actually work together.

For what you are asking, to hash an entire email (or even a non-empty piece of it), `TMessage` would have to be saved in its entirety to a `TStream` first, then that `TStream` hashed and transmitted as is, where the hash is either injected into the TStream beforehand, or the `TStream` is split into 2 pieces so the 1st piece can be transmitted followed by the hash followed by the remainder of the `TStream`.  But in any case, this is simply not how `TIdMessageClient` is designed to send a `TIdMessage`, and that design can't be changed under the current architecture, which is why I'm now tagging this ticket for consideration for Indy 12 (Indy 11 is going to be a maintenance release, no new features).

**Comment by VBGAMER45** [comment]
I was able to find a way to get DKIM working with multipart messages with an external library libdkim. Required altering IdMessage.pas the TIdMIMEBoundaryStrings.GenerateBoundary generation caused an issue since it was regenerated anytime  i saved to a stream so had to write a property to set a custom boundary. I did use memorystream to fetch the message for dkim signing.

**Comment by rlebeau** [comment]
An issue to watch out for is SMTP dot transparency (which requires escaping leading dots with double-dots).  DKIM needs to be applied to a pre-escaped email.  But `TIdMessage` always dot-escapes an email when saving it to a `TStream` or preparing it for sending over a socket, regardless of whether the email is being used for SMTP or not (see https://github.com/IndySockets/Indy/issues/135).  There is a workaround for that - the `TIdMessageHelper` class in the `IdMessageHelper.pas` unit, which adds a `AUseDotTransparency` parameter to `TIdMessage.SaveTo...()` and `TIdMessage.LoadFrom...()` methods, which you can set to False.  But then you run into the issue that `TIdSMTP` is not designed to send streams, so you would have to issue your own SMTP `DATA` command and dump the modified `TStream` directly to the socket, escaping it manually as needed.
